sdk/storage/azure-storage-blob/azure/storage/blob/_download.py

```diff
@@ -531,81 +529,112 @@ def chunks(self):
+        errors. If size/chars is unspecified or negative all bytes will be read.
 
-        :param Optional[int] size:
+        :param int size:
             The number of bytes to download from the stream. Leave unspecified
-            or set to -1 to download all bytes.
+            or set negative to download all bytes.
+        :param Optional[int] chars:
+            The number of chars to download from the stream. Leave unspecified
+            or set negative to download all chars. Note, this can only be used
+            when encoding is specified on `download_blob`.
         :returns:
             The requested data as bytes or a string if encoding was specified. If
             the return value is empty, there is no more data to read.
         :rtype: T
         """
-        if size == -1:
-            return self.readall()
         # Empty blob or already read to the end
         if size == 0 or self._offset >= self.size:
             return b'' if not self._encoding else ''
 
-        stream = BytesIO()
-        remaining_size = size
-
-        # Start by reading from current_content if there is data left
-        if self._offset < len(self._current_content):
-            start = self._offset
-            length = min(remaining_size, len(self._current_content) - self._offset)
-            read = stream.write(self._current_content[start:start + length])
-
-            remaining_size -= read
-            self._offset += read
-            if self._progress_hook:
-                self._progress_hook(self._offset, self.size)
-
-        if remaining_size > 0:
-            start_range = self._get_downloader_start_with_offset()
-
-            # End is the min between the remaining size, the file size, and the end of the specified range
-            end_range = min(start_range + remaining_size, self._file_size)
-            if self._end_range is not None:
-                end_range = min(end_range, self._end_range + 1)
+        if size > -1 and self._encoding:
+            warnings.warn(
+                "Size parameter specified with text encoding enabled. It is recommended to use chars "
+                "to read a specific number of characters instead."
+            )
+        if size > -1 and chars is not None:
+            raise ValueError("Cannot specify both size and chars.")
+        if not self._encoding and chars is not None:
+            raise ValueError("Must specify encoding to read chars.")
+        if self._text_mode and size > -1:
+            raise ValueError("Stream has been partially read in text mode. Please use chars.")
+
+        download_complete = self._download_offset >= self.size
+        if not self._text_mode and chars is not None:
+            self._text_mode = True
+            self._decoder = codecs.getincrementaldecoder(self._encoding)('strict')
+            self._current_content = self._decoder.decode(self._current_content, final=download_complete)
```

Comment:
So if `final` is `False` here - does that mean it will only decode what it can? What happens to any left over bytes?

Reply:
Yes, it will decode and return what it can, and any extra bytes will stay buffered in the decoder object (this is a stateful decoder). And only when `final=True` will it raise a `UnicodeError` if there are bytes left over.